File: scala/simple_text_editor/simple_text_editor_13.scala
// Ahoy, matey! This be a simple text editor program, designed to help ye navigate the treacherous seas of text manipulation.
// Ye'll be able to add, delete, and undo text changes with the grace of a seasoned sailor. Hoist the mainsail and let's set sail!

import scala.collection.mutable.Stack

// The global treasure chest where all the text be stored
var text: String = ""
val undoStack: Stack[String] = Stack()

// Function to add text to the editor
def addText(newText: String): Unit = {
    undoStack.push(text)
    text += newText
}

// Function to delete text from the editor
def deleteText(count: Int): Unit = {
    undoStack.push(text)
    text = text.dropRight(count)
}

// Function to undo the last action
def undo(): Unit = {
    if (undoStack.nonEmpty) {
        text = undoStack.pop()
    }
}

// Function to display the current text
def displayText(): Unit = {
    println(s"Current text: $text")
}

// Function to change the weather (not really, just a fun name for a function)
def changeWeather(weather: String): Unit = {
    println(s"The weather be $weather today!")
}

// Main function to run the editor
def main(args: Array[String]): Unit = {
    // Adding some text to start with
    addText("Ahoy, matey! ")
    displayText()

    // Adding more text
    addText("Welcome to the text editor. ")
    displayText()

    // Deleting some text
    deleteText(8)
    displayText()

    // Undoing the last action
    undo()
    displayText()

    // Changing the weather for fun
    changeWeather("sunny")
}

File: swift/simple_text_editor/simple_text_editor_13.swift
//
// Ahoy, matey! Welcome to the Simple Text Editor!
// This here program be designed to help ye navigate the treacherous waters of text editing.
// With this fine piece of code, ye'll be able to add, remove, and display text like a true buccaneer!
// Prepare yerself for a journey filled with verbose comments and a plethora of variables!
//

import Foundation

// The grand treasure chest where all text be stored
var treasureChest: [String] = []

// Function to add text to the treasure chest
func addText(_ text: String) {
    // Hoist the sails and add the text to the chest
    treasureChest.append(text)
}

// Function to remove text from the treasure chest
func removeText(at index: Int) {
    // Beware, ye scallywag! Make sure the index be within the bounds of the chest
    if index >= 0 && index < treasureChest.count {
        // Cast the text overboard
        treasureChest.remove(at: index)
    }
}

// Function to display all the text in the treasure chest
func displayText() {
    // Unfurl the map and reveal the contents of the chest
    for (index, text) in treasureChest.enumerated() {
        print("\(index): \(text)")
    }
}

// Function to count the number of texts in the treasure chest
func countTexts() -> Int {
    // Count the pieces of eight in the chest
    return treasureChest.count
}

// Function to clear all the text from the treasure chest
func clearAllText() {
    // Swab the deck and clear the chest
    treasureChest.removeAll()
}

// Function to edit text in the treasure chest
func editText(at index: Int, with newText: String) {
    // Replace the old text with the new text if the index be valid
    if index >= 0 && index < treasureChest.count {
        treasureChest[index] = newText
    }
}

// Function to search for text in the treasure chest
func searchText(_ query: String) -> [Int] {
    // Search the seven seas for the query and return the indices
    var foundIndices: [Int] = []
    for (index, text) in treasureChest.enumerated() {
        if text.contains(query) {
            foundIndices.append(index)
        }
    }
    return foundIndices
}

// Function to display the weather (not really needed, but here it be)
func displayWeather() {
    let weather = "Sunny"
    print("The weather be \(weather) today!")
}

// Main function to run the text editor
func main() {
    // Set sail and start the text editor
    addText("Ahoy, matey!")
    addText("Shiver me timbers!")
    displayText()
    removeText(at: 1)
    displayText()
    editText(at: 0, with: "Yo ho ho!")
    displayText()
    let searchResults = searchText("Yo")
    print("Found at indices: \(searchResults)")
    clearAllText()
    displayText()
    displayWeather()
}

// Call the main function to start the adventure
main()

